dbms/src/Server/CertificateReloader.cpp

```diff
@@ -0,0 +1,61 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Server/CertificateReloader.h>
+
+#if Poco_NetSSL_FOUND
+
+#include <Poco/Net/Utility.h>
+
+namespace DB
+{
+namespace
+{
+/// Call set process for certificate.
+int callSetCertificate(SSL * ssl, void * arg)
+{
+    return CertificateReloader::instance().setCertificate(ssl, arg);
+}
+} // namespace
+
+/// This is callback for OpenSSL. It will be called on every connection to obtain a certificate and private key.
+int CertificateReloader::setCertificate(SSL * ssl, void * arg)
+{
+    auto * context = static_cast<Context *>(arg);
+    auto security_config = context->getSecurityConfig();
+    auto [cert_path, key_path] = security_config->getCertAndKeyPath();
+    LOG_DEBUG(log, "setCertificate callback called, cert_path: {}, key_path: {}", cert_path, key_path);
+    Poco::Crypto::X509Certificate cert(cert_path);
+    Poco::Crypto::EVPPKey key("", key_path);
+    SSL_use_certificate(ssl, const_cast<X509 *>(cert.certificate()));
+    SSL_use_PrivateKey(ssl, const_cast<EVP_PKEY *>(static_cast<const EVP_PKEY *>(key)));
```

Comment:
Shall we check the return values of these two calls?

Reply:
> Shall we check the return values of these two calls?

Got, I will check the returen val.

Reply:
> > Shall we check the return values of these two calls?
> 
> Got, I will check the returen val.

And IMO, the check below can ensure the error handling safe, but will lose some debug info.

Reply:
> > Shall we check the return values of these two calls?
> 
> Got, I will check the returen val.

And currently when cert or key is invalid, the following log will be printed:
```
[2022/11/24 08:33:24.548 +08:00] [ERROR] [Exception.cpp:89] ["Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = Cannot read file: Faild to load certificate from: /data1/zzywq/tiup_deploy/dev1/tiflash-6319/tls_old/tiflash.crt, e.what() = Cannot read file"] [source=ServerErrorHandler] [thread_id=58]
[2022/11/24 08:33:54.550 +08:00] [ERROR] [Exception.cpp:89] ["Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = OpenSSLException: EVPKey::loadKey(string): error:06000066:public key routines:OPENSSL_internal:DECODE_ERROR: error:0900000c:PEM routines:OPENSSL_internal:ASN.1 encoding routines: error:00000000:invalid library (0):OPENSSL_internal:invalid library (0), e.what() = OpenSSLException"] [source=ServerErrorHandler] [thread_id=58]
```

Reply:
> > > Shall we check the return values of these two calls?
> > 
> > 
> > Got, I will check the returen val.
> 
> And currently when cert or key is invalid, the following log will be printed:
> 
> ```
> [2022/11/24 08:33:24.548 +08:00] [ERROR] [Exception.cpp:89] ["Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = Cannot read file: Faild to load certificate from: /data1/zzywq/tiup_deploy/dev1/tiflash-6319/tls_old/tiflash.crt, e.what() = Cannot read file"] [source=ServerErrorHandler] [thread_id=58]
> [2022/11/24 08:33:54.550 +08:00] [ERROR] [Exception.cpp:89] ["Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = OpenSSLException: EVPKey::loadKey(string): error:06000066:public key routines:OPENSSL_internal:DECODE_ERROR: error:0900000c:PEM routines:OPENSSL_internal:ASN.1 encoding routines: error:00000000:invalid library (0):OPENSSL_internal:invalid library (0), e.what() = OpenSSLException"] [source=ServerErrorHandler] [thread_id=58]
> ```

After I check the return val, the log is like:
```
[2022/11/24 08:39:24.553 +08:00] [ERROR] [Exception.cpp:89] ["Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = OpenSSLException: EVPKey::loadKey(string): error:06000066:public key routines:OPENSSL_internal:DECODE_ERROR: error:0900000c:PEM routines:OPENSSL_internal:ASN.1 encoding routines: error:00000000:invalid library (0):OPENSSL_internal:invalid library (0), e.what() = OpenSSLException"] [source=ServerErrorHandler] [thread_id=57]
[2022/11/24 08:39:54.548 +08:00] [ERROR] [Exception.cpp:89] ["Poco::Exception. Code: 1000, e.code() = 0, e.displayText() = Cannot read file: Faild to load certificate from: /data1/zzywq/tiup_deploy/dev1/tiflash-6319/tls_old/tiflash.crt, e.what() = Cannot read file"] [source=ServerErrorHandler] [thread_id=57]
```
No difference since ssl is not null and if the cert or key is invalid, Poco will raise exception.
But error checking is always necessary.